File: src/Day04.kt
fun main() {
    val input = readInput("day04.txt")

    val maxX = input.size
    val maxY = input[0].length

    var result = 0
    for (needle in listOf("XMAS", "SAMX")) {
        for (i in 0 until maxX) {
            for (j in 0 until maxY) {
                // Right
                var matched = 4
                for (k in needle.indices) {
                    if (j + k >= maxX || input[i][j + k] != needle[k]) {
                        matched -= 1
                        break
                    }
                }
                // Down
                for (k in needle.indices) {
                    if (i + k >= maxY || input[i + k][j] != needle[k]) {
                        matched -= 1
                        break
                    }
                }
                // Down-right
                for (k in needle.indices) {
                    if (i + k >= maxY || j + k >= maxX || input[i + k][j + k] != needle[k]) {
                        matched -= 1
                        break
                    }
                }
                // Down-left
                for (k in needle.indices) {
                    if (i + k >= maxY || j - k < 0 || input[i + k][j - k] != needle[k]) {
                        matched -= 1
                        break
                    }
                }

                result += matched
            }
        }
    }

    println(result)

    // Part 2
    result = 0
    val a = input
    val n = setOf('M', 'S')
    for (i in 1 until maxX - 1) {
        for (j in 1 until maxY - 1) {
            if (a[i][j] != 'A') {
                continue
            }

            val lu = a[i - 1][j - 1]
            val ll = a[i + 1][j - 1]
            val ru = a[i - 1][j + 1]
            val rl = a[i + 1][j + 1]

            if (setOf(lu, rl) != n || setOf(ru, ll) != n) {
                continue
            }

            result += 1
        }
    }
    println(result)
}

File: src/day04.rs
use std::collections::HashSet;
use crate::read_lines;

pub fn solve() {
    let input: Vec<_> = read_lines("day04.txt");

    let mut result = 0;
    let needle = "XMAS";

    let max_y = input.len() as i32;
    let max_x = input[0].len() as i32;

    // Part 1
    for y in 0..max_y {
        for x in 0..max_x {
            // Let's scan in all 8 directions at once
            for d1 in -1i32..=1i32 {
                for d2 in -1i32..=1i32 {
                    if d1 == 0 && d2 == 0 { // No direction, not interesting
                        continue;
                    }

                    let mut found = true;
                    for i in 0i32..needle.len() as i32 {
                        // I.e. (0; 1) means "scanning right"
                        // (1, 1) means "scanning down-right" etc.
                        let yk = y + i * d1;
                        let xk = x + i * d2;
                        if yk < 0 || yk >= max_y || xk < 0 || xk >= max_x {
                            found = false;
                            break;
                        }
                        let n = needle.chars().nth(i as usize).unwrap();
                        let st = input[yk as usize].chars().nth(xk as usize).unwrap();
                        if st != n {
                            found = false;
                            break;
                        }
                    }
                    if found {
                        result += 1;
                    }
                }
            }
        }
    }

    println!("{}", result);

    // Part 2
    result = 0;
    let golden = HashSet::from(['M', 'S']);
    for y in 1..(max_y as usize - 1)  {
        for x in 1..(max_x as usize - 1) {
            let curr = input[y].chars().nth(x).unwrap();
            if curr != 'A' {
                continue
            }
            let lu = input[y - 1].chars().nth(x - 1).unwrap();
            let rl = input[y + 1].chars().nth(x + 1).unwrap();

            let ll = input[y + 1].chars().nth(x - 1).unwrap();
            let ru = input[y - 1].chars().nth(x + 1).unwrap();
            if golden.eq(&HashSet::from([lu, rl])) && golden.eq(&HashSet::from([ll, ru])) {
                result += 1;
            }
        }
    }
    println!("{}", result)
}
